gpu4pyscf/lib/ecp/ecp_type1_ip.cu

```diff
@@ -14,6 +14,116 @@
  * limitations under the License.
  */
 
+template <int orderi, int orderj, int LI, int LJ> __device__
+void type1_cart_unrolled_kernel(double *gctr,
+                const int ish, const int jsh, const int ksh,
+                const int *ecpbas, const int *ecploc,
+                const int *atm, const int *bas, const double *env)
+{
+    const int npi = bas[NPRIM_OF+ish*BAS_SLOTS];
+    const int npj = bas[NPRIM_OF+jsh*BAS_SLOTS];
+    const double *ai = env + bas[PTR_EXP+ish*BAS_SLOTS];
+    const double *aj = env + bas[PTR_EXP+jsh*BAS_SLOTS];
+    const double *ci = env + bas[PTR_COEFF+ish*BAS_SLOTS];
+    const double *cj = env + bas[PTR_COEFF+jsh*BAS_SLOTS];
+    const double *ri = env + atm[PTR_COORD+bas[ATOM_OF+ish*BAS_SLOTS]*ATM_SLOTS];
+    const double *rj = env + atm[PTR_COORD+bas[ATOM_OF+jsh*BAS_SLOTS]*ATM_SLOTS];
+
+    const int atm_id = ecpbas[ATOM_OF+ecploc[ksh]*BAS_SLOTS];
+    const double *rc = env + atm[PTR_COORD+atm_id*ATM_SLOTS];
+
+    double rca[3], rcb[3];
+    rca[0] = rc[0] - ri[0];
+    rca[1] = rc[1] - ri[1];
+    rca[2] = rc[2] - ri[2];
+    rcb[0] = rc[0] - rj[0];
+    rcb[1] = rc[1] - rj[1];
+    rcb[2] = rc[2] - rj[2];
+    const double r2ca = rca[0]*rca[0] + rca[1]*rca[1] + rca[2]*rca[2];
+    const double r2cb = rcb[0]*rcb[0] + rcb[1]*rcb[1] + rcb[2]*rcb[2];
+
+    double ur = 0.0;
+    for (int kbas = ecploc[ksh]; kbas < ecploc[ksh+1]; kbas++){
+        ur += rad_part(kbas, ecpbas, env);
+    }
+
+    constexpr int LIJ1 = LI+LJ+1;
+    constexpr int LIJ3 = LIJ1*LIJ1*LIJ1;
+
+    __shared__ double rad_ang[LIJ3];
+    set_shared_memory(rad_ang, LIJ3);
+
+    const double fac = 16.0 * M_PI * M_PI * _common_fac[LI] * _common_fac[LJ];
+    for (int ip = 0; ip < npi; ip++){
+        for (int jp = 0; jp < npj; jp++){
+            double rij[3];
+            double ai_prim = ai[ip];
+            double aj_prim = aj[jp];
+            rij[0] = ai_prim * rca[0] + aj_prim * rcb[0];
+            rij[1] = ai_prim * rca[1] + aj_prim * rcb[1];
+            rij[2] = ai_prim * rca[2] + aj_prim * rcb[2];
+            const double k = 2.0 * norm3d(rij[0], rij[1], rij[2]);
+            const double aij = ai_prim + aj_prim;
+            
+            __shared__ double rad_all[LIJ1*LIJ1];
+            type1_rad_part(rad_all, LI+LJ, k, aij, ur);
+
+            const double eij = exp(-ai_prim*r2ca - aj_prim*r2cb);
+            const double eaij = eij * pow(-2.0*ai_prim, orderi) * pow(-2.0*aj_prim, orderj);
+            const double ceij = eaij * ci[ip] * cj[jp];
+            type1_rad_ang<LI+LJ>(rad_ang, rij, rad_all, fac*ceij);
+            //type1_rad_ang(rad_ang, LI+LJ, rij, rad_all, fac*ceij);
+            __syncthreads();
+        }
+    }
+
+    constexpr int NFI = (LI+1)*(LI+2)/2;
+    constexpr int NFJ = (LJ+1)*(LJ+2)/2;
+    double fi[3*NFI];
```

Comment:
Is shared memory enough to store fi and fj?

Reply:
They are probably not in register. Some addressing operations are called in the next loop bloc, like `double* fx_i = fi + (ix+1)*ix/2;` . These addresses cannot be determined at the compile stage. Compiler has to put them in the local memory.